packages/web/src/components/basic/StateProvider.js

```diff
@@ -43,39 +47,60 @@ const filterByKeys = (state, allowedKeys) =>
 		}),
 		{},
 	);
-class StateProvider extends Component {
-	constructor(props) {
-		super(props);
-		this.state = {
-			searchState: filterByKeys(getSearchState(filterProps(props)), props.includeKeys),
-		};
-	}
-	static getDerivedStateFromProps(props) {
-		return {
-			searchState: filterByKeys(getSearchState(filterProps(props)), props.includeKeys),
-		};
-	}
-	isStateChanged(prevState, nextState) {
-		return JSON.stringify(nextState) !== JSON.stringify(prevState);
-	}
-	shouldComponentUpdate(nextProps, nextState) {
-		// Only apply when componentIds is defined
-		if (!nextProps.strict || this.isStateChanged(this.state, nextState)) {
-			return true;
-		}
-		return false;
-	}
-	componentDidUpdate(prevProps, prevState) {
-		const { onChange } = this.props;
-		if (onChange && this.isStateChanged(prevState, this.state)) {
-			onChange(prevState.searchState, this.state.searchState);
-		}
-	}
-	render() {
-		const { searchState } = this.state;
-		return getComponent({ searchState }, this.props);
-	}
+
+function usePrevious(value) {
+	const ref = useRef();
+	useEffect(() => {
+		ref.current = value;
+	});
+	return ref.current;
 }
+
+function StateProvider(props) {
+	const populateReduxState = useSelector(state => ({
+		selectedValues: filterByComponentIds(state.selectedValues, props),
+		queryLog: filterByComponentIds(state.queryLog, props),
+		dependencyTree: filterByComponentIds(state.dependencyTree, props),
+		componentProps: filterByComponentIds(state.props, props),
+		hits: filterByComponentIds(state.hits, props),
+		aggregations: filterByComponentIds(state.aggregations, props),
+		isLoading: filterByComponentIds(state.isLoading, props),
+		error: filterByComponentIds(state.error, props),
+		promotedResults: filterByComponentIds(state.promotedResults, props),
+	}));
+
+	const mergeProps = { ...props, ...populateReduxState };
+
+	const searchState = filterByKeys(
+		getSearchState(filterProps(mergeProps)),
+		mergeProps.includeKeys,
+	);
+
+	// preserve searchState's previous state and call it in comparision check
+	const prevState = usePrevious(searchState);
+
+	const stateChangeComparision = JSON.stringify(prevState) !== JSON.stringify(searchState);
+	const shouldUpdate = useMemo(() => !props.strict || stateChangeComparision, [
+		stateChangeComparision,
+	]);
+
+	// prevent extra re-render on change of every respective prop if value remains same towards comparision
+	const view = useMemo(
+		() => <React.Fragment>{getComponent({ searchState }, mergeProps)}</React.Fragment>,
+		[shouldUpdate],
+	);
+
+	// // Listen for state change and trigger the callback for useEffect
+	React.useEffect(() => {
+		// console.log('Redux state change', prevState);
+
+		const { onChange } = props;
+		// call onChange() if it has been defined in parent
+		onChange && onChange(prevState.searchState, this.state.searchState);
```

Comment:
`this.state.searchState` is wrongly used over here. Since it is a functional component it should be only `searchState`

Reply:
Since we are changing state over here, we should add a memory leak check.

example:

```
useEffect(() => {
   let isMounted = true;
   
  if (isMounted) {
    // on chnage logic
  }
  
 return function cleanUp() {
   isMounted = false;
 }
}, [])

```

This will help us prevent any memory leak when a state is change is executed on unmounted component.

Reply:
So should replace it with this example code or add it below the onchange useeffect hook?